### PR TITLE
Merge barebones frontend/backend for stripe payment

### PR DESCRIPTION
**Pull request by haoyangw**
Changes:
- Add create stripe session endpoint so that frontend can be redirected to stripe API for payment
- Add skeleton webhook endpoint to connect with Stripe API to test payment events
    - None of the event handlers currently do anything to avoid making changes to backend state until schemas are finalised
- Create basic billing page on frontend that uses new create session backend endpint to perform stripe payment
- At the moment payment success does NOT update user tier, so this is not production ready

### CODE REVIEW

**Comment by haoyangw** [comment]
Nevermind, I just learnt that I can forward stripe webhooks to a local deployment for testing before 'going live', let me hold on to this pull request until the endpoint is actually tested

**Comment by seelengxd** [comment]
note: missing files with endpoint

**Comment by haoyangw** [comment]
> note: missing files with endpoint

Apologies, just committed the actual endpoint code

File: backend/src/subscriptions/dependencies.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+def retrieve_subscription(
+    id: int,
+    user=Depends(get_current_user),
+    session=Depends(get_session),
+) -> Subscription:
+    subscription = session.scalar(
+        select(Subscription)
+        .where(Subscription.id == id)
+        .options(
+            selectinload(Subscription.user.and_(Subscription.user_id == user.id)),
```
Comment (seelengxd) [inline backend/src/subscriptions/dependencies.py:19]:
```suggestion
            selectinload(Subscription.user),
```

not necessary, when you saw that line its usually for cases like event has 10 likes (user actually not part of the relationship), only one of them belong the user and i don't want to show the other 9

in this case user is part of the relationship, and that would be part of the join condition already

File: backend/src/subscriptions/models.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+class Subscription(Base):
+    __tablename__ = "subscription"
+
+    id: Mapped[int] = mapped_column(primary_key=True)
+    # Make user_id nullable since the frontend needs to match subscription.id
+    #   to user.id for us, which might only happen after stripe webhook is triggered
+    #   and this table is populated
+    user_id: Mapped[int] = mapped_column(ForeignKey("user.id"), nullable=True)
+    price_id: Mapped[str]
```
Comment (seelengxd) [inline backend/src/subscriptions/models.py:25]: how do you plan to link price_id to tier?

File: backend/src/subscriptions/dependencies.py (hunk 1 above)
Comment (haoyangw) [reply backend/src/subscriptions/dependencies.py]: Thanks! have made the suggested change

**Comment by haoyangw** [comment]
TODO:
- Handle payment failures and subscription cancel and other failure cases

Right now, only the ideal case(successful payment) is implemented and handled

**Comment by haoyangw** [comment]
Progress update:
- Subscription cancel is now tested and working
- Stripe customer portal is now accessible from user billing page for changing payment method and cancelling subscription

**Pull request merged**